### ISSUE TITLE
Where to upload runtime?

### ISSUE THREAD

**Issue by sultanshakir** [rrainn/aws-lambda-custom-node-runtime#5]
I have tried to zip and upload the files generated by this utility in several ways and keep getting errors.  Where does the node-v8.10.0-linux-x64 directory go?

`START RequestId: 5efac149-762e-4720-9356-7a38be0ff0c1 Version: $LATEST
/opt/bootstrap: line 3: ./node-v8.10.0-linux-x64/bin/node: No such file or directory
END RequestId: 5efac149-762e-4720-9356-7a38be0ff0c1
REPORT RequestId: 5efac149-762e-4720-9356-7a38be0ff0c1	Duration: 58.63 ms	Billed Duration: 100 ms	Memory Size: 128 MB	Max Memory Used: 6 MB	
RequestId: 5efac149-762e-4720-9356-7a38be0ff0c1 Error: Runtime exited with error: exit status 127
Runtime.ExitError
`

**Comment by fishcharlie** [comment]
@sultanshakir Was able to get this working. Here is what I did:

1. Create a new directory
2. `cd` into that directory
3. Run `aws-lambda-custom-node-runtime <node_version>` (ex. `aws-lambda-custom-node-runtime 11.3.0`)
4. Create file `index.js` with the code below
5. Open directory that you originally created
6. Zip all files within the direction into one zip (don't zip the directory but the files within the directory you created, should be 4 folders/files in the example I'm doing)
7. Create new Lambda function with runtime set to `Provide your own bootstrap`
8. Upload zip file as the code for that Lambda function
9. Update handler to `index.handler`
10. Run Lambda function
11. See output in logs of `Hello World`

```js
exports.handler = () => {
    console.log("Hello World");
};
```

So in short, it should just live alongside your code.

```
charliefish@Charlies-MacBook-Pro tmp3 % ls
bootstrap		node-v11.3.0-linux-x64
index.js		node_runtime.js
```

**Comment by f100024** [comment]
@fishcharlie thank you for your answer. I've followed your steps but got error
```

events.js:301
    throw new TypeError('"listener" argument must be a function');
    ^

TypeError: "listener" argument must be a function
    at ClientRequest.once (events.js:301:11)
    at new ClientRequest (_http_client.js:77:10)
    at Object.exports.request (http.js:31:10)
    at request (/var/task/node_runtime.js:68:17)
    at run (/var/task/node_runtime.js:5:2)
    at Object.<anonymous> (/var/task/node_runtime.js:48:1)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
```

zipped folder
```
bootstrap  index.js  node_runtime.js  node-v6.10.3-linux-x64
```

index.js
```
exports.handler = () => {
    console.log("Hello World");
};
```
Could you please explain what I'm doing wrong?

**UPD:** It works with node 11.3.0 But exit with next as describes by @sultanshakir:
```
 Error: Runtime exited without providing a reason
Runtime.ExitError
```

**Comment by fishcharlie** [comment]
@f100024 So in the call stack it mentions `node_runtime.js:68:17`. Which refers to:

https://github.com/rrainn/aws-lambda-custom-node-runtime/blob/09076f1662591a5a7b5814b2922b3ea3c8c89dff/templates/node_runtime.js#L68

It looks like between Node v6 and the latest version they changed that to not allow for 3 parameters, but only 2. The solution here might be to refactor that code to check to see what version of Node.js is being run maybe?? And change the parameters passed in based on that. We'd have to find the exact version that change was made tho, and it does add complexity, which kinda sucks.

https://nodejs.org/docs/latest-v6.x/api/http.html#http_http_request_options_callback

Feel free to try to come up with a solution and submit a pull request.

---

As for `Error: Runtime exited without providing a reason` on Node 11.3.0, I have no idea why that would be occurring. That isn't the same error as @sultanshakir had tho, because yours is exiting without a reason.

That is honestly an **awful** error by AWS Lambda. Gives no detail and doesn't help debug at all. For all we know that could be an internal issue with AWS Lambda outside of our control.

If I were you, I'd probably contact AWS Support and see if they can't provide more details about why that is occurring or help give more information to debug better.

**Comment by f100024** [comment]
Ok, many thanks!
